File: backend/src/main/java/com/ucaacp/backend/controller/PsychKnowledgeController.java
package com.ucaacp.backend.controller;

import com.ucaacp.backend.annotation.CheckLogin;
import com.ucaacp.backend.annotation.CheckUserRole;
import com.ucaacp.backend.entity.DTO.PsychKnowledgeDTO;
import com.ucaacp.backend.entity.PsychKnowledge;
import com.ucaacp.backend.entity.PsychKnowledgeReport;
import com.ucaacp.backend.entity.enums.UserRole;
import com.ucaacp.backend.service.PsychKnowledgeService;
import com.ucaacp.backend.utils.return_object.ReturnObject;
import jakarta.servlet.http.HttpSession;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/psych_knowledge")
@CrossOrigin(origins = "http://localhost:3000") // 开发环境使用
public class PsychKnowledgeController {

    @Autowired
    private PsychKnowledgeService psychKnowledgeService;

    @CheckLogin
    @GetMapping("/list_public")
    public ReturnObject listPublic(HttpSession session) {
        List<PsychKnowledgeDTO> knowledgeDTOList=psychKnowledgeService.findAllPassedPsychKnowledgeDTO();
        if(knowledgeDTOList!=null){
            return ReturnObject.success(knowledgeDTOList);
        }else{
            return ReturnObject.fail("获取心理知识科普失败");
        }
    }



    @CheckLogin
    @PostMapping("/report")
    public ReturnObject report(@RequestBody Map<String,Object> reportRequest,HttpSession session) {

        Integer knowledgeId=(Integer)reportRequest.get("knowledgeId");
        String reportReason=(String)reportRequest.get("reportReason");
        String reporterUsername=(String)reportRequest.get("reporterUsername");

        PsychKnowledgeReport psychKnowledgeReport=new PsychKnowledgeReport();
        psychKnowledgeReport.setKnowledgeId(knowledgeId);
        psychKnowledgeReport.setReportReason(reportReason);
        psychKnowledgeReport.setReporterUsername(reporterUsername);

        if(psychKnowledgeService.report(psychKnowledgeReport)!=null){
            return ReturnObject.success();
        }else{
            return ReturnObject.fail("举报失败");
        }
    }

    @CheckLogin
    @GetMapping("/teacher/mine")
    public ReturnObject teacherMine(@RequestParam Map<String,Object> params,HttpSession session) {
        String teacherUsername=(String)params.get("teacherUsername");
        List<PsychKnowledgeDTO> psychKnowledgeDTOList=psychKnowledgeService.findPsychKnowledgeReportDTOByTeacher(teacherUsername);
        if(psychKnowledgeDTOList!=null){
            return ReturnObject.success(psychKnowledgeDTOList);
        }else{
            return ReturnObject.fail("获取心理知识科普失败");
        }
    }


    @CheckLogin
    @CheckUserRole(UserRole.TEACHER)
    @PostMapping("/teacher/post")
    public ReturnObject teacherPost(@RequestBody Map<String,Object> postRequest,HttpSession session) {
        String teacherUsername=(String)postRequest.get("teacherPublisherUsername");
        String title=(String)postRequest.get("title");
        String content=(String)postRequest.get("content");

        PsychKnowledge psychKnowledge=new PsychKnowledge(title,content,teacherUsername);

        if(psychKnowledgeService.post(psychKnowledge)!=null){
            return ReturnObject.success();
        }else{
            return ReturnObject.fail("提交失败");
        }

    }

    @CheckLogin
    @CheckUserRole(UserRole.TEACHER)
    @PostMapping("/teacher/invoke")
    public ReturnObject teacherInvoke(@RequestBody Map<String,Object> invokeRequest,HttpSession session) {
        Integer knowledgeId=(Integer)invokeRequest.get("knowledgeId");
        if(psychKnowledgeService.invoke(knowledgeId)>=0){
            return ReturnObject.success();
        }else{
            return ReturnObject.fail("撤回失败");
        }
    }

}

File: frontend/src/controller/PsychKnowledgeController.ts
import {Controller} from "./Controller";
import {PsychKnowledgeDTO} from "../entity/DTO/PsychKnowledgeDTO";
import {PsychKnowledgeReport} from "../entity/PsychKnowledgeReport";



export interface PsychKnowledgeReportRequest {
    knowledgeId:number;
    reportReason:string;
    reporterUsername:string;
}

export interface PsychKnowledgePostRequest {
    title:string;
    content:string;
    teacherPublisherUsername:string;
}

export interface PsychKnowledgePassRequest{
    knowledgeId:number;
    adminReviewerUsername:string;
}

export interface PsychKnowledgeBanRequest{
    knowledgeId:number;
    adminReviewerUsername:string;
}



export class PsychKnowledgeController extends Controller{
    //列出所有已通过审核的科普
    listPublic=this._get<null,PsychKnowledgeDTO[]>("/api/psych_knowledge/list_public");
    //推荐一个科普
    recommend=this._get<null,PsychKnowledgeDTO>("/api/psych_knowledge/recommend");
    //举报心理知识科普
    report=this._post<PsychKnowledgeReportRequest,any>("api/psych_knowledge/report");
    //列出教师发布的心理知识科普
    teacherMine=this._get<{teacherUsername:string},PsychKnowledgeDTO[]>("api/psych_knowledge/teacher/mine");
    //教师提交科普
    teacherPost=this._post<PsychKnowledgePostRequest,any>("api/psych_knowledge/teacher/post");
    //教师撤回科普
    teacherInvoke=this._post<{knowledgeId:number},any>("api/psych_knowledge/teacher/invoke");
    //列出管理员审核的心理知识科普
    adminReviewed=this._get<{adminReviewerUsername:string},any>("api/psych_knowledge/admin/reviewed");
    //管理员通过心理知识科普
    adminPass=this._post<PsychKnowledgePassRequest,any>("api/psych_knowledge/admin/pass");
    //管理员审核心理知识科普
    adminBan=this._post<PsychKnowledgeBanRequest,any>("api/psych_knowledge/admin/ban");
    //获取所有未处理的心理知识科普
    pending=this._get<null,PsychKnowledgeDTO[]>("/api/psych_knowledge/pending");
    //获取所有被举报的心理知识科普
    reported=this._get<null,PsychKnowledgeDTO[]>("/api/psych_knowledge/reported");
    //获取某科普的所有举报列表
    listReport=this._get<{knowledgeId:number},PsychKnowledgeReport[]>("api/psych_knowledge/list_report");
    //删除某条举报
    adminReportDelete=this._post<{reportId:number},any>("api/psych_knowledge/admin/report/delete")
}